src/analysis/lattices/lift.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_analysis_lattices_lift_h
+#define wasm_analysis_lattices_lift_h
+
+#include <optional>
+#include <utility>
+
+#include "../lattice.h"
+#include "bool.h"
+
+namespace wasm::analysis {
+
+template<Lattice L> struct Lift {
+  struct Element : std::optional<typename L::Element> {
+    bool isBottom() const noexcept { return !this->has_value(); }
+    bool operator==(const Element& other) const noexcept {
+      return (isBottom() && other.isBottom()) ||
+             (!isBottom() && !other.isBottom() && **this == *other);
```

Comment:
Why wouldn't` **this == *other` work if both are bottom?

Reply:
We can't evaluate `**this` if `isBottom` because that would be dereferencing an empty optional.

Reply:
Ah, I see, then the new bottom element is basically the "nothing" case of the optional? So the extra bit of the optional marks that new item, basically. That's what I was missing.

At the risk of seeming repetitive: comments :smiley: